h/traversal/group_membership.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import dataclass
+
+from pyramid.httpexceptions import HTTPNotFound
+
+from h.exceptions import InvalidUserId
+from h.models import Group, GroupMembership, User
+
+
+@dataclass
+class GroupMembershipContext:
+    group: Group
```

Comment:
Not a suggestion, just checking my understanding of the context here:

Could we have 


```
class GroupMembershipContext:
￼    target_group: Group
     target_user: User
     target_membership: GroupMembership | None
     
     autheticated_user: User
     autheticaed_user_membership: GroupMembership | None
```

or those last two are just duplicating the info over identity and making the context less reusable?

Reply:
I think it's wrong to include information about the authenticated user in the context: `request.identity` is about the authenticated user, whereas `context` is about the target resource(s). For example h has this [`identity_permits(identity, context, permission)`](https://github.com/hypothesis/h/blob/9a762648ae633117ffd6114d0dd2a6e953232429/h/security/permits.py#L9-L31) function and you can do things like this:

```python
context = FooContext()
for user in users:
    identity = Identity.from_models(user)
    if identity_permits(identity, context, "foo_permission"):
        pass
```

...that is, you can construct a single `context` object and use it repeatedly to ask whether different identities have a given permission in that context.

I haven't tried it but I think you could do a similar thing in pure Pyramid (without h's custom `identity_permits()` function) by constructing requests:

```python
context = FooContext()
for user in users:
    request = Request(...)
    if request.has_permission(context, "foo_permission"):
        pass
```

...but it's more difficult because you need to somehow create a `request` object that's authenticated as the user in question. May need to use [`paster_bootstrap()`](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/paster.html#pyramid.paster.bootstrap)